libraries/CurieBLE/src/internal/BLEServiceImp.cpp

```diff
@@ -347,44 +370,143 @@ uint8_t BLEServiceImp::discoverResponseProc(bt_conn_t *conn,
-        BLECharacteristicImp* chrcCurImp = NULL;
-        BLECharacteristicNodePtr node = chrcHeader->next;
         
-        pr_debug(LOG_MODULE_BLE, "%s-%d: node-%p",__FUNCTION__, __LINE__, node);
-        // Discover next service
-        while (node != NULL)
+        if (false == _reading)
         {
-            chrcCurImp = node->value;
-            
-            if (NULL == _cur_discover_chrc)
+            discoverNextCharacteristic(device);
+        }
+    }
+    return retVal;
+}
+
+void BLEServiceImp::discoverNextCharacteristic(BLEDevice &bledevice)
+{
+    const BLECharacteristicLinkNodeHeader* chrcHeader = &_characteristics_header;
+    BLECharacteristicImp* chrcCurImp = NULL;
+    BLECharacteristicNodePtr node = chrcHeader->next;
+    
+    //pr_debug(LOG_MODULE_BLE, "%s-%d: node-%p",__FUNCTION__, __LINE__, node);
+    // Discover next service
+    while (node != NULL)
+    {
+        chrcCurImp = node->value;
+        
+        if (NULL == _cur_discover_chrc)
+        {
+            bool result = chrcCurImp->discoverAttributes(&bledevice);
+            pr_debug(LOG_MODULE_BLE, "%s-%d",__FUNCTION__, __LINE__);
+            if (result == true)
             {
-                bool result = chrcCurImp->discoverAttributes(&device);
-                pr_debug(LOG_MODULE_BLE, "%s-%d",__FUNCTION__, __LINE__);
-                if (result == true)
-                {
-                    // Record the current discovering service
-                    _cur_discover_chrc = chrcCurImp;
-                    break;
-                }
+                // Record the current discovering service
+                _cur_discover_chrc = chrcCurImp;
+                break;
             }
-            else if (_cur_discover_chrc == chrcCurImp)
+        }
+        else if (_cur_discover_chrc == chrcCurImp)
+        {
+            // Find next discoverable service
+            _cur_discover_chrc = NULL;
+        }
+        node = node->next;
+    }
+}
+
+bool BLEServiceImp::readCharacteristic(const BLEDevice &bledevice, uint16_t handle)
+{
+    int retval = 0;
+    bt_conn_t* conn = NULL;
+    
+    if (true == BLEUtils::isLocalBLE(bledevice))
+    {
+        // GATT server can't write
+        return false;
+    }
+    
+    if (_reading)
+    {
+        return false;
+    }
+    
+    _read_params.func = profile_characteristic_read_rsp_process;
+    _read_params.handle_count = 1;
+    _read_params.single.handle = handle - 1;
+    _read_params.single.offset = 0;
+    
+    if (0 == _read_params.single.handle)
+    {
+        // Discover not complete
+        return false;
+    }
+    
+    conn = bt_conn_lookup_addr_le(bledevice.bt_le_address());
+    if (NULL == conn)
+    {
+        return false;
+    }
+    // Send read request
+    retval = bt_gatt_read(conn, &_read_params);
+    bt_conn_unref(conn);
+    if (0 == retval)
+    {
+        _reading = true;
+    }
+    pr_debug(LOG_MODULE_BLE, "%s-%d", __FUNCTION__, __LINE__);
```

Comment:
I believe all these calls to pr_debug will incur some serial I/O. If this is the case, please remove them..